### PR TITLE
Revert "REF: back IntervalArray by a single ndarray (#37047)"

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
This reverts commit 9cb372376fc78aa66e2559de919592007b74cfaa.

See discussion in https://github.com/pandas-dev/pandas/pull/37047/

@jbrockmendel can you give this a check? (there were already several conflicts)

### CODE REVIEW

**Comment by jbrockmendel** [comment]
I'll take a look, tentatively looks OK.

By funny coincidence, I _just_ got the first of the promised perf improvements working.

File: pandas/core/arrays/interval.py [hunk 1]
```diff
@@ -210,22 +192,83 @@ def __new__(
...
+            # GH 19262: dtype must be an IntervalDtype to override inferred
+            dtype = pandas_dtype(dtype)
+            if not is_interval_dtype(dtype):
+                msg = f"dtype must be an IntervalDtype, got {dtype}"
+                raise TypeError(msg)
+            elif dtype.subtype is not None:
+                left = left.astype(dtype.subtype)
+                right = right.astype(dtype.subtype)
+
+        # coerce dtypes to match if needed
```
Comment (jreback) [inline pandas/core/arrays/interval.py:224]: can you factor this out into a function (maybe_cast_inputs), this is crazy otherwise

**Review by jreback — CHANGES_REQUESTED** [review]
ok with the change back.

File: pandas/core/arrays/interval.py (hunk 1 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/interval.py:224]: This is how it was before (I am just reverting / fixing conflicts). I prefer to keep this PR a clean revert, and do any other refactor as follow-up

Comment (jreback) [reply pandas/core/arrays/interval.py:224]: i c ok.

**Comment by jreback** [comment]
@jbrockmendel ok here?

**Review by jreback — APPROVED** [review]

**Comment by jbrockmendel** [comment]
looks fine

**Comment by jorisvandenbossche** [comment]
>  I just got the first of the promised perf improvements working.

Do you have a branch for it that you can push? I am curious to see whether the principles of the speed-up also couldn't be applied to the "2 1D-arrays" model

**Comment by jbrockmendel** [comment]
> Do you have a branch for it that you can push? I am curious to see whether the principles of the speed-up also couldn't be applied to the "2 1D-arrays" model

shortly, i hope.  The approach is to do `self._combined.view("complex128")` and then do set operations on that to avoid casting to object.

**Comment by jorisvandenbossche** [comment]
I can imagine that an additional copy to create the combined array could still be worth it compared to casting to object

**Pull request merged**